[FLINK-3030][web frontend] Enhance dashboard to show execution attempts

mushketyk opened the pull request:
I've added support for displaying execution attempts in the dashboard. 

The JIRA item also includes that we should add accumulators for different attempts, but it seems that accumulators are going to be deprecated (judging from this JIRA item https://issues.apache.org/jira/browse/FLINK-4527) so I didn't implement that.

Here are few screenshots of the new feature:
A list of opened attempts are collapsed:

![attempts collapsed](https://cloud.githubusercontent.com/assets/592286/18147690/c287947e-6fcd-11e6-8968-ff47283635b7.png)

A list of opened attempts is opened:

![opened attempts](https://cloud.githubusercontent.com/assets/592286/18147738/e9e3d8d4-6fcd-11e6-94c1-00036df3348b.png)

If there is no attempts the button is not displayed:

![no attempts](https://cloud.githubusercontent.com/assets/592286/18147715/d5251c1e-6fcd-11e6-9b95-b72d80c05051.png)
- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)
- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added
- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

StephanEwen commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/HandlersUtils.java:
```diff
@@ -0,0 +1,85 @@
+import java.io.IOException;
+import java.util.Map;
+
+public final class HandlersUtils {
```
How about giving this a more revealing name. Like `JsonGenerators`?
You could also simply move the `serializeExecutionAttempt(...)` method as a static method to the `SubtaskExecutionAttemptDetailsHandler`.

mushketyk replied:
Good point. I would prefer to have a separate class `JsonGenerators` that will have just one responsibility.

StephanEwen commented on flink-runtime-web/web-dashboard/web/css/index.css:
```diff
@@ -228,7 +228,6 @@
   overflow: hidden;
   padding: 2px 4px;
   font-size: 14px;
-  -webkit-border-radius: 2px;
```
How did this come to be? Are you using a different version of glup or the dependencies to build this?

mushketyk replied:
Good catch. I'll take a look at what `gulp` version am I using and check how I can fix it.
Could you please check what version of `gulp` are you using? I don't think it is documented anywhere.

StephanEwen commented:
Good change, all in all, good work!

Few minor comments, otherwise +1

mushketyk commented:
Thank you for the review @StephanEwen
I'll update the PR in the evening.

mushketyk commented:
@StephanEwen I've renamed the class as you suggested.
Regarding the CSS change, I don't know why these changes are added. I have the required (according to README.md) `gulp` version:
```
> $ gulp -version 
[21:06:35] CLI version 3.9.1
[21:06:35] Local version 3.9.1

```

I wonder if a `.styl` file was changed in one of the previous commits but it was not compiled to `.css`.

mushketyk commented:
@StephanEwen Should I change anything else in this PR?

mushketyk commented:
Is there anything else that should be updated in this PR? It has been in review for more than a month.

StephanEwen commented:
Good to go from my side.
@rmetzger and @iampeter - since you are working on the web UI currently, can you try and merge this?

zentol commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JsonGenerators.java:
```diff
@@ -0,0 +1,86 @@
+			readBytes = (LongCounter) metrics.get(AccumulatorRegistry.Metric.NUM_BYTES_IN);
+			writeBytes = (LongCounter) metrics.get(AccumulatorRegistry.Metric.NUM_BYTES_OUT);
+			readRecords = (LongCounter) metrics.get(AccumulatorRegistry.Metric.NUM_RECORDS_IN);
+			writeRecords = (LongCounter) metrics.get(AccumulatorRegistry.Metric.NUM_RECORDS_OUT);
```
that is correct. It also has to be rebased due to the introduction of the ArchivedExecutionGraph.

mushketyk replied:
Ok. I'll rebase the change and fix merge conflicts.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java:
```diff
@@ -241,6 +241,18 @@ public Execution getPriorExecutionAttempt(int attemptNumber) {
 		}
 	}
 
+	public Execution getExecutionAttempt(int attemptNumber) {
```
i would prefer if you would instead use getCurrentExecutionAttempt if attemptNumber==0 and getPriorExecutionAttempt if attemptNumber>0, for the sole purpose of not having to many methods in the ExecutionVertex class.

mushketyk replied:
Ok. Will do this.

iampeter commented:
@rmetzger will you be merging? if so, `flink-runtime-web/web-dashboard/web/js/index.js` just needs a `gulp`

rmetzger commented:
Okay, I'll run gulp when merging.

mushketyk commented:
@rmetzger @iampeter 
I've rebased the code and updated it according to your suggestions.

uce commented:
Really like this! I will check it out and try to play around with it a little and merge if everything works as expected. How does this interplay with https://issues.apache.org/jira/browse/FLINK-5645 @zentol? The `SubtaskAllExecutionAttemptsHandler` uses `execAttempt.getIOMetrics()`.

---

In general, I think the web frontend requires a general overhaul making execution attempts and failures first class citizens. The current state of this won't scale well to long running deployments where potentially many failures happen over time. Then you have to scroll to each subtask and find the failed attempt you are looking for etc.

uce commented:
It worked well, but I'm a little hesitant to merge it as it is. Running with higher parallelism than 1 I think that the button is pretty invasive:

![screen shot 2017-01-27 at 11 50 37](https://cloud.githubusercontent.com/assets/1756620/22368508/239f1768-e487-11e6-996a-a56aa89ab922.png)

Can we make this a little less eye-catching? Maybe just a link next to the attempt number.

mushketyk commented:
Hi @uce 

Thank you for your review.
I'll update the UI in the next couple of days.

mushketyk commented:
Hi @uce,

I've rebased changes and updated the UI according to your suggestion. Now it looks like:
<img width="1389" alt="new attempts link" src="https://cloud.githubusercontent.com/assets/592286/22408112/2165724a-e66b-11e6-8008-6042a97ea412.png">

What do you think about this?

aljoscha commented:
I'm closing this as "Abandoned", since there is no more activity and the code base has moved on quite a bit. Please re-open this if you feel otherwise and work should continue.

There is a new dashboard implementation now.
